### PR TITLE
Add Stop Loss Take Gain

### PR DESCRIPTION
**Pull request by imadelh**
Here is the strategy discussed. I tried to explain the logic in `enter_exit_loss_gain`. 
Will add unittests soon after confirming that the goal of the strategy is what we are looking for.

### CODE REVIEW

File: src/backlight/strategies/exit.py [hunk 1]
```diff
@@ -36,6 +36,45 @@ def _exit_transaction(
     return Transaction(timestamp=exit_index, amount=-trade.sum())
 
 
+def _exit_transaction_gain_loss(
+    df: pd.DataFrame,
+    trade: pd.Series,
+    loss_threshold: float,  # Negative
```
Comment (keisuke-umezawa) [inline src/backlight/strategies/exit.py]: To align with other functions, please use positive value.
https://github.com/AlpacaDB/backlight/blob/master/src/backlight/strategies/exit.py#L234-L235

File: src/backlight/strategies/exit.py [hunk 2]
```diff
@@ -36,6 +36,45 @@ def _exit_transaction(
...
+    gain_threshold: float,  # Positive
+) -> Transaction:
+
+    # Get an exit indice that satisfies gain or loss limit
+    current_signal = df["pred"][0]
+    if df["pred"][0] == 1.0:  # If signal UP, we buy at ask, sell at bid
```
Comment (keisuke-umezawa) [inline src/backlight/strategies/exit.py]: For the first step, how about just using `mid` price? In that case, we can write logic simply. Or, we can write unittest, otherwise it is quite difficult to judge the algorithm is fine.

Comment (keisuke-umezawa) [reply src/backlight/strategies/exit.py]: When using `mid`, logic should be:

```
def exit_at_loss_and_gain(
    mkt: MarketData,
    entries: Trades,
    max_holding_time: pd.Timedelta,
    loss_threshold: float,
    gain_threshold: float,
) :
    assert loss_threshold >= 0.0
    assert gain_threshold >= 0.0

    def _exit_at_loss_and_gain(df: pd.DataFrame, trade: pd.Series) -> pd.Series:
        prices = df.mid

        amount = trade.sum()
        entry_price = prices.iloc[0]
        pl_per_amount = np.sign(amount) * (prices - entry_price)
        is_stop_loss = pl_per_amount <= -initial_stop
        is_take_gain = pl_per_amount >= gain_threshold

        return is_initial_stop | is_take_gain

    return exit_by_max_holding_time(mkt, None, entries, max_holding_time, _exit_at_loss_and_gain)
```

Comment (keisuke-umezawa) [reply src/backlight/strategies/exit.py]: If we want to ask/bid pricing for loss cut, 
I think it is better to create functions that will replace `df.mid`.

```
prices = calculate_price(mkt, sig)

def calculate_price(mkt, sig):
    if sig == UP:
        return mkt.ask
    if sig == DOWN:
        return mkt.bid
    return mkt.mid
```

Comment (imadelh) [reply src/backlight/strategies/exit.py]: Ok I got it. I just updated it to take mid price for now.

File: src/backlight/strategies/exit.py [hunk 3]
```diff
@@ -256,3 +256,30 @@ def _exit_by_trailing_stop(df: pd.DataFrame, trade: pd.Series) -> pd.Series:
         return is_initial_stop | is_trailing_stop
 
     return exit(mkt, None, entries, _exit_by_trailing_stop)
+
+
+def exit_at_loss_at_gain(
```
Comment (keisuke-umezawa) [inline src/backlight/strategies/exit.py]: `exit_at_loss_and_gain` is better?

**Comment by keisuke-umezawa** [comment]
And I also want to see the result using this functionality. If you have some analysis, please open PR in forecast and hold a discussion!

File: src/backlight/strategies/exit.py (hunk 3 above)
Comment (imadelh) [reply src/backlight/strategies/exit.py]: changed

File: tests/strategies/test_exit.py [hunk 4]
```diff
@@ -185,3 +185,39 @@ def test_exit_by_trailing_stop(market, signal, entries):
     )
 
     pd.testing.assert_frame_equal(trades, expected)
+
+
+def test_exit_at_loss_and_gain(market, signal, entries):
+    max_holding_time = pd.Timedelta("3min")
+    gain_threshold = 100
```
Comment (keisuke-umezawa) [inline tests/strategies/test_exit.py]: Please add another test when gain_threshold is small.

Comment (imadelh) [reply tests/strategies/test_exit.py]: Sure, I'll add it.

**Comment by imadelh** [comment]
I'll also run it for dumped signals as discussed with a grid for confidence levels vs loss levels on forecast.

File: tests/strategies/test_exit.py [hunk 5]
```diff
@@ -187,37 +187,72 @@ def test_exit_by_trailing_stop(market, signal, entries):
...
+                [-2.0],
+                [-1.0],
+                [-1.0],
+                [1.0],
+                [1.0],
+                [1.0],
+                [1.0],
+                [0.0],
+                [0.0],
+                [-1.0],
```
Comment (keisuke-umezawa) [inline tests/strategies/test_exit.py]: Did you check that the logic is correct by unittest? In that case, you can merge it.

Comment (imadelh) [reply tests/strategies/test_exit.py]: I double checked and it seems that logic is followed by unittests. I'll meerge it
